webapp/lib/view.dart

```diff
@@ -500,161 +475,186 @@ class BatchRepliesConfigurationView extends PackageConfiguratorView {
         ..append(addsTagsContainer)
     );
   }
+}
+
+class TagListView extends BaseView {
+  List<TagView> tagElements;
+  DivElement _tagsContainer;
+  SpanElement _tagsActionContainer;
+  Function onTagChangedCallback;
 
-  SpanElement _addTagAction(List<String> tags) {
-    return new SpanElement()
+  TagListView(this.tagElements, Map<String, model.TagStyle> tagList, this.onTagChangedCallback, [bool tagsEditable = false]) {
+    _tagsContainer = new DivElement()
+      ..classes.add('tags');
+    _tagsActionContainer = new SpanElement()
       ..classes.add('tags__actions')
       ..append(
         new ButtonElement()
           ..classes.add('button-add-tag')
           ..text = '+'
           ..onClick.listen((event) {
-            var tagActions = (event.target as Element).parent;
-            var tagList = tagActions.parent;
-            _addTagDropdown(tagActions, tagList, model.tags);
+            if (tagsEditable) {
+              _tagsActionContainer.insertAdjacentElement('beforebegin', new TagView('', model.TagStyle.Normal, onTagChangedCallback, tagsEditable).renderElement);
+            } else {
+              _addTagDropdown(tagList, onTagChangedCallback);
+            }
             event.stopPropagation();
           })
       );
+    _tagsContainer.append(_tagsActionContainer);
+    tagElements.forEach((tag) {
+      _tagsActionContainer.insertAdjacentElement('beforebegin', tag.renderElement);
+    });
   }
 
-  void _addTagDropdown(Element tagActionsContainer, Element tagListContainer, List<String> tags) {
-    document.querySelectorAll('.add-tag-dropdown').forEach((dropdown) => dropdown.remove());
+  DivElement get renderElement => _tagsContainer;
+
+  void _addTagDropdown(Map<String, model.TagStyle> tags, Function(String, model.TagStyle) onTagChangedCallback) {
     var tagListDropdown = new Element.ul()
       ..classes.add('add-tag-dropdown');
-    tagActionsContainer.append(tagListDropdown);
-    var tagsToShow = tags.isEmpty ? ['--None--'] : tags;
+    var tagsToShow = tags.isEmpty ? ['--None--'] : tags.keys;
     for (var tag in tagsToShow) {
       tagListDropdown.append(
         new Element.li()
           ..classes.add('add-tag-dropdown__item')
           ..text = tag
           ..onClick.listen((event) {
-            if (tag != '--None--') _addTag((event.target as Element).text, tagListContainer);
+            if (tag != '--None--') onTagChangedCallback(tag, tags[tag]);
```

Comment:
prefer early exit for the special case rather than conditional general case:
```suggestion
            if (tag == '--None--') return;
            onTagChangedCallback(tag, tags[tag]);
```